src/index/Index.Text.cpp

```diff
@@ -10,23 +12,109 @@
 #include "../util/Simple8bCode.h"
 #include "./FTSAlgorithms.h"
 #include "./Index.h"
 
+// _____________________________________________________________________________
+cppcoro::generator<ContextFileParser::Line> Index::contextFileLines(
+    const std::string& contextFile, bool useRdfVocabularyAsText) {
+  auto localeManager = _textVocab.getLocaleManager();
+  // ROUND 1: Read from context file aka wordsfile if not empty. Remember
+  // the last context id for the (optional) second round.
+  //
+  // TODO: For now, EITHER do this round OR the other, see the respective
+  // comment in `addTextFromContextFile` below.
+  TextRecordIndex contextId = TextRecordIndex::make(0);
+  if (!useRdfVocabularyAsText) {
+    ContextFileParser::Line line;
+    ContextFileParser p(contextFile, localeManager);
+    ad_utility::HashSet<string> items;
+    while (p.getLine(line)) {
+      contextId = line._contextId;
+      co_yield line;
+    }
+    if (contextId > TextRecordIndex::make(0)) {
+      contextId = contextId.incremented();
+    }
+  }
+  // ROUND 2: Consider RDF vocabulary as the "contexts" aka text records.
+  if (useRdfVocabularyAsText) {
+    auto isWordChar = [](char c) -> bool { return std::isalnum(c); };
+    for (VocabIndex index = VocabIndex::make(0); index.get() < _vocab.size();
+         index = index.incremented()) {
+      auto text = _vocab.at(index);
+      if (isLiteral(text)) {
+        ContextFileParser::Line entityLine{text, true, contextId, 1};
+        co_yield entityLine;
+        size_t pos = 0;
+        size_t posEnd = text.rfind('"');
+        if (posEnd == std::string::npos) posEnd = text.size();
+        while (pos < posEnd) {
+          while (pos < posEnd && !isWordChar(text[pos])) ++pos;
+          size_t posWordBegin = pos;
+          while (pos < posEnd && isWordChar(text[pos])) ++pos;
+          if (pos > posWordBegin) {
+            auto word = localeManager.getLowercaseUtf8(
+                text.substr(posWordBegin, pos - posWordBegin));
+            ContextFileParser::Line wordLine{word, false, contextId, 1};
+            co_yield wordLine;
+          }
+        }
+        contextId = contextId.incremented();
+      }
+    }
+  }
+}
+
 // _____________________________________________________________________________
 void Index::addTextFromContextFile(const string& contextFile) {
+  LOG(INFO) << std::endl;
+  LOG(INFO) << "Adding text index ..." << std::endl;
   string indexFilename = _onDiskBase + ".text.index";
-  size_t nofLines = passContextFileForVocabulary(contextFile);
+  // If `contextFile` ends with ".vocabulary", don't actually read from a file,
+  // but instead consider the IRIs and literals from the RDF vocabulary as text
+  // records.
+  //
+  // TODO: The generator `contextFileLines` above can actually both read from a
+  // file and then also consider the RDF vocabulary, we just need to add a
+  // proper command-line option to enable this. In the meantime, it's either-or.
```

Comment:
Now I understand. I already have a PR lying around that implements cleaner command-line parsing for the index-builder, maybe when we merge this first it will be easier to add command-line options.

Reply:
I now added a proper command-line option already to this version, it was easy enough